IdleTimeout with Jetty HTTP/2 and InputStreamResponseListener

Here is the exception
```
   [junit4]   2> Caused by: java.util.concurrent.TimeoutException: idle_timeout
   [junit4]   2> 	at org.eclipse.jetty.http2.client.http.SolrConnection.onIdleTimeout(SolrConnection.java:143) ~[java/:?]
   [junit4]   2> 	at org.eclipse.jetty.http2.client.http.SolrTransportHttp2$SessionListenerPromise.onIdleTimeout(SolrTransportHttp2.java:260) ~[java/:?]
   [junit4]   2> 	at org.eclipse.jetty.http2.HTTP2Session.notifyIdleTimeout(HTTP2Session.java:1166) ~[http2-common-9.4.14.v20181114.jar:9.4.14.v20181114]
   [junit4]   2> 	at org.eclipse.jetty.http2.HTTP2Session.onIdleTimeout(HTTP2Session.java:1004) ~[http2-common-9.4.14.v20181114.jar:9.4.14.v20181114]
   [junit4]   2> 	at org.eclipse.jetty.http2.HTTP2Connection.onIdleExpired(HTTP2Connection.java:151) ~[http2-common-9.4.14.v20181114.jar:9.4.14.v20181114]
   [junit4]   2> 	at org.eclipse.jetty.io.AbstractEndPoint.onIdleExpired(AbstractEndPoint.java:401) ~[jetty-io-9.4.14.v20181114.jar:9.4.14.v20181114]
   [junit4]   2> 	at org.eclipse.jetty.io.IdleTimeout.checkIdleTimeout(IdleTimeout.java:166) ~[jetty-io-9.4.14.v20181114.jar:9.4.14.v20181114]
   [junit4]   2> 	at org.eclipse.jetty.io.IdleTimeout$1.run(IdleTimeout.java:50) ~[jetty-io-9.4.14.v20181114.jar:9.4.14.v20181114]
   [junit4]   2> 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[?:?]
   [junit4]   2> 	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
   [junit4]   2> 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[?:?]
```

After digging into the Jetty code, here is what I found
* The sever did successfully write the response to the client.
* The client did receive the data but the listener is not called with new data.
* At first I thought that `HeadersFrame.endStream=true`  -> subsequent data does not trigger the listener. But it is not true `HeadersFrame.endStream=false`
* The reason here is `HttpReceiverOverHTTP2` is reused between multiple req-resp calls. For some reason the `Receiver` does not reset and stuck at SUCCEEDED state. 

So here is the flow that leads to the problem

1. `HttpReceiverOverHTTP2.ContentNotifier.process()` is called, `prevDataInfo.frame` have `endStream=true` therefore the method return `Action.SUCCEEDED`.  `IteratingCallback` then get changed to `SUCCEEDED` and the `HttpExchange.response` state is changed to `TERMINATED`. This whole process is called in a thread that is reading data from `InputStreamResponseListener`. In this case `HttpReceiverOverHTTP2.reset()` is called, note that the `HttpExchange.Response` state before this reset call is PENDING and after the response state is TERMINATED.
2. In another thread of `HttpClient` executor. The same `HttpReceiverOverHTTP2.ContentNotifier.process()` is called with same `prevDataInfo.frame` (same frame with same hashCode()). But since the exchange response state is TERMINATED. `HttpReceiverOverHTTP2.reset()` is not called.
3. For the next request response that use same `HttpReceiverOverHTTP2`, even data is offered to the queue, the Receiver do nothing since `ContentNotifier` state now is `SUCCEEDED`

--- a/jetty-http2/http2-http-client-transport/src/main/java/org/eclipse/jetty/http2/client/http/HttpConnectionOverHTTP2.java
+++ b/jetty-http2/http2-http-client-transport/src/main/java/org/eclipse/jetty/http2/client/http/HttpConnectionOverHTTP2.java
@@ -51,6 +51,7 @@ public class HttpConnectionOverHTTP2 extends HttpConnection implements Sweeper.S
     private final AtomicBoolean closed = new AtomicBoolean();
     private final AtomicInteger sweeps = new AtomicInteger();
     private final Session session;
+    private boolean recycleHttpChannels;
 
     public HttpConnectionOverHTTP2(HttpDestination destination, Session session)
     {
@@ -61,6 +62,16 @@ public class HttpConnectionOverHTTP2 extends HttpConnection implements Sweeper.S
     public Session getSession()
     {
         return session;
+    }
+
+    public boolean isRecycleHttpChannels()
+    {
+        return recycleHttpChannels;
+    }
+
+    public void setRecycleHttpChannels(boolean recycleHttpChannels)
+    {
+        this.recycleHttpChannels = recycleHttpChannels;
     }
 
     @Override
@@ -99,7 +110,7 @@ public class HttpConnectionOverHTTP2 extends HttpConnection implements Sweeper.S
             // Recycle only non-failed channels.
             if (channel.isFailed())
                 channel.destroy();
-            else
+            else if (isRecycleHttpChannels())
                 idleChannels.offer(channel);
         }
         else

--- a/jetty-http2/http2-http-client-transport/src/main/java/org/eclipse/jetty/http2/client/http/HttpReceiverOverHTTP2.java
+++ b/jetty-http2/http2-http-client-transport/src/main/java/org/eclipse/jetty/http2/client/http/HttpReceiverOverHTTP2.java
@@ -213,21 +213,21 @@ public class HttpReceiverOverHTTP2 extends HttpReceiver implements Stream.Listen
         @Override
         protected Action process()
         {
-            DataInfo dataInfo;
+            if (dataInfo != null)
+            {
+                dataInfo.callback.succeeded();
+                if (dataInfo.frame.isEndStream())
+                    return Action.SUCCEEDED;
+            }
+
             synchronized (this)
             {
                 dataInfo = queue.poll();
             }
 
             if (dataInfo == null)
-            {
-                DataInfo prevDataInfo = this.dataInfo;
-                if (prevDataInfo != null && prevDataInfo.frame.isEndStream())
-                    return Action.SUCCEEDED;
                 return Action.IDLE;
-            }
 
-            this.dataInfo = dataInfo;
             ByteBuffer buffer = dataInfo.frame.getData();
             if (buffer.hasRemaining())
                 responseContent(dataInfo.exchange, buffer, this);
@@ -245,13 +245,6 @@ public class HttpReceiverOverHTTP2 extends HttpReceiver implements Stream.Listen
         }
 
         @Override
-        public void succeeded()
-        {
-            dataInfo.callback.succeeded();
-            super.succeeded();
-        }
-
-        @Override
         protected void onCompleteSuccess()
         {
             responseSuccess(dataInfo.exchange);
@@ -262,6 +255,14 @@ public class HttpReceiverOverHTTP2 extends HttpReceiver implements Stream.Listen
         {
             dataInfo.callback.failed(failure);
             responseFailure(failure);
+        }
+
+        @Override
+        public boolean reset()
+        {
+            queue.clear();
+            dataInfo = null;
+            return super.reset();
         }
     }
 

